File: Medium_Solutions/49 Group Anagrams .py
class Solution:

    def groupAnagrams(self, strs):
        
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s)==len(t):
            Hash1={}
            Hash2={}
            for i in range(0,len(s)):
               
                if  s[i] in Hash1:
                    Hash1[s[i]]+=1
                    
                else:
                    Hash1[s[i]]=1
                        
                if t[i] in Hash2:
                    Hash2[t[i]]+=1 
                
                else:
                    Hash2[t[i]]=1
                
 
            Hash1 = dict(sorted(Hash1.items()))
            Hash2 = dict(sorted(Hash2.items()))
            
            if Hash1==Hash2:
                return True
            else: 
                return False
        else:
            return False        

            
            
        
S1=Solution()
print(S1.groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

File: Medium_Solutions/array_and_hash_map/49_group_anagrams.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
  vector<vector<string>> groupAnagrams(vector<string> &strs) {
    unordered_map<string, vector<string>> res;
    for (const string &str : strs) {
      vector<int> count(26, 0);
      for (char ch : str)
        count[ch - 'a']++;
      string key = to_string(count[0]);
      for (int index = 1; index < 26; ++index)
        key += ',' + to_string(count[index]);
      res[key].push_back(str);
    }
    vector<vector<string>> result;
    for (const pair<string, vector<string>> &p : res)
      result.push_back(p.second);
    return result;
  }
};

int main() {
  Solution sol;
  vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};

  vector<vector<string>> grouped = sol.groupAnagrams(strs);

  cout << "[\n";
  for (auto &group : grouped) {
    cout << "  [ ";
    for (auto &word : group)
      cout << "\"" << word << "\" ";
    cout << "]\n";
  }
  cout << "]\n";

  return 0;
}
